livesync_citadel/indico_livesync_citadel/schemas.py

```diff
@@ -0,0 +1,245 @@
+    else:
+        principals = obj.get_access_list()
+
+    acl = set(itertools.chain.from_iterable(_get_identifiers(x) for x in principals))
+    if not len(acl):
+        acl.add('')
+    return {'read': sorted(acl), 'owner': [''], 'update': [''], 'delete': ['']}
+
+
+def _get_attachment_content(attachment):
+    if attachment.type == AttachmentType.file:
+        from indico_livesync_citadel.plugin import LiveSyncCitadelPlugin
+        parsedfile = parser.from_file(attachment.absolute_download_url,
+                                      LiveSyncCitadelPlugin.settings.get('tika_server'))['content']
+        return parsedfile
+
+
+def _get_attachment_contributionid(attachment):
+    return attachment.folder.contribution.id if attachment.folder.link_type == LinkType.contribution else None
+
+
+def _get_attachment_subcontributionid(attachment):
+    return attachment.folder.subcontribution.id if attachment.folder.link_type == LinkType.subcontribution else None
+
+
+def _get_people_list(obj):
+    return [
+        '{} ({})'.format(pl.full_name, pl.affiliation) if pl.affiliation else pl.full_name
+        for pl in obj.person_links
+    ]
+
+
+def _get_event_data(event):
+    id = event.id
+    category_path = _get_category_path(event)
+    event_type = event.type
+    title = event.title
+    description = unicode(event.description)
+    location = _get_location(event)
+    speakers_chairs = _get_people_list(event)
+    url = event.external_url
+    return {'id': id, 'category_path': category_path, 'event_type': event_type, 'title': title,
+            'description': description, 'location': location, 'speakers_chairs': speakers_chairs, 'url': url}
+
+
+class EventSchema(mm.Schema):
+    _access = fields.Function(_get_event_acl)
+    _data = fields.Function(_get_event_data)
+    creation_date = fields.DateTime(attribute='created_dt', format='%Y-%m-%dT%H:%M')
+    start_date = fields.DateTime(attribute='start_dt', format='%Y-%m-%dT%H:%M')
+    end_date = fields.DateTime(attribute='end_dt', format='%Y-%m-%dT%H:%M')
+
+
+def _get_attachment_data(attachment):
+    id = attachment.id
+    category_path = _get_category_path(attachment)
+    event_id = attachment.folder.event.id
+    contribution_id = _get_attachment_contributionid(attachment)
+    subcontribution_id = _get_attachment_subcontributionid(attachment)
+    filename = attachment.file.filename
+    content = _get_attachment_content(attachment)
+    url = attachment.absolute_download_url
+
+    return {'id': id, 'category_path': category_path, 'event_id': event_id, 'contribution_id': contribution_id,
+            'subcontribution_id': subcontribution_id, 'filename': filename, 'content': content, 'url': url}
+
+
+class AttachmentSchema(mm.Schema):
+    _access = fields.Function(_get_attachment_acl)
+    _data = fields.Function(_get_attachment_data)
+    creation_date = fields.DateTime(attribute='modified_dt', format='%Y-%m-%dT%H:%M')
+
+
+def _get_contribution_data(contribution):
```

Comment:
This looks like something that should be in a nested schema. But I don't know how that would work since we probably pass the contribution itself to the schema's `dump()`...

Reply:
Probably the best solution is using this schema for the main contribution data and using a `post_dump` decorator to wrap it with the extra data:

https://marshmallow.readthedocs.io/en/stable/marshmallow.decorators.html